src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+
+            var expressionType = semanticDocument.SemanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (expressionType is null or IErrorTypeSymbol)
+            {
+                return null;
+            }
+
+            var (title, actions) = AddActions(semanticDocument, expression, cancellationToken);
+
+            if (actions.Length > 0)
+            {
+                return new CodeActionWithNestedActions(title, actions, isInlinable: true);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and refactors all the call sites
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForRefactoringAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            var updatedCallSitesSolution = await RewriteCallSitesAsync(document, expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+            return updatedCallSitesSolution;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and a new method that calls upon the updated method so that all
+        /// refactorings are coalesced to one location
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForTrampolineAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, bool trampoline, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+            var updatedInvocationsSolution = await RewriteCallSitesWithNewMethodAsync(document,
+                expression, methodSymbolInfo, allOccurrences, parameterName, methodCallSites, trampoline, cancellationToken).ConfigureAwait(false);
+            return updatedInvocationsSolution;
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (IsExpressionRemovable(document, expression))
+            {
+                var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+                return syntaxFacts.GetIdentifierOfVariableDeclarator(expression.Parent!.Parent!).ValueText;
```

Comment:
Just something to consider, but this line could be rolled into `IsExpressionRemovable`, and that could be renamed `TryGetVariableDeclaratorName` or something, and then you wouldn't need the exclamation marks. Until someone looks into IsExpressionRemovable it looks like a very bold assumption here.

Reply:
I now see `IsExpressionRemovable` is used later on, so perhaps ignore me.

Reply:
That's true, I can change that.

Reply:
Yeah I am still wondering if there's a nicer way to do this, or if I just put a comment indicating that it is checked in "IsExpressionRemovable" and that should be enough?